src/main/scala/io/iohk/ethereum/db/storage/ReferenceCountNodeStorage.scala

```diff
@@ -36,99 +36,106 @@ class ReferenceCountNodeStorage(nodeStorage: NodeStorage, pruningOffset: BigInt,
-    * marked as unused in a certain block number tag, might be used later)
+    * Looks for the StoredNode snapshots based on block number and saves (or deletes) them
     *
-    * @param blockNumber
-    * @param nodeStorage
-    * @return
+    * @param blockNumber BlockNumber to rollback
+    * @param nodeStorage NodeStorage
     */
-  private def prune(blockNumber: BigInt, nodeStorage: NodeStorage): Int = {
-
-    val key = pruneKey(blockNumber)
-
-    nodeStorage.get(key)
-      .map(pruneCandidatesFromBytes)
-      .map { pruneCandidates: PruneCandidates =>
-        // Get Node from storage and filter ones which have references = 0 now (maybe they were added again after blockNumber)
-        val pruneCandidateNodes = pruneCandidates.nodeKeys.map(nodeStorage.get)
-        val pruneCandidateNodesWithKeys = pruneCandidateNodes.zip(pruneCandidates.nodeKeys)
-        pruneCandidateNodesWithKeys.collect {
-          case (Some(candidate), candidateKey) if storedNodeFromBytes(candidate).references == 0 => candidateKey }
-      }.map(nodeKeysToDelete => {
-        // Update nodestorage removing all nodes that are no longer being used
-        nodeStorage.update(key +: nodeKeysToDelete, Nil)
-        nodeKeysToDelete.size
-      }).getOrElse(0)
+  override def rollback(blockNumber: BigInt, nodeStorage: NodeStorage): Unit =
+    withSnapshotCount(blockNumber, nodeStorage) { (snapshotsCountKey, snapshotCount) =>
+      // Get all the snapshots
+      val snapshots = snapshotKeysUpTo(blockNumber, snapshotCount)
+        .flatMap(key => nodeStorage.get(key).map(snapshotFromBytes))
+
+      // Transform them to db operations
+      val (toRemove, toUpsert) = snapshots.foldLeft(Seq.empty[NodeHash], Seq.empty[(NodeHash, NodeEncoded)]) {
+        // Undo Actions
+        case ((r, u), StoredNodeSnapshot(nodeHash, Some(sn))) => (r, (nodeHash -> storedNodeToBytes(sn)) +: u)
+        case ((r, u), StoredNodeSnapshot(nodeHash, None)) => (nodeHash +: r, u)
+      }
+      // also remove snapshot as we have done a rollback
+      nodeStorage.update(toRemove :+ snapshotsCountKey, toUpsert)
+    }
+
+  private def withSnapshotCount(blockNumber: BigInt, nodeStorage: NodeStorage)(f: (ByteString, BigInt) => Unit) = {
+    val snapshotsCountKey = getSnapshotsCountKey(blockNumber)
+    // Look for snapshot count for given block number
+    val maybeSnapshotCount = nodeStorage.get(snapshotsCountKey).map(snapshotsCountFromBytes)
+    maybeSnapshotCount match {
+      case Some(snapshotCount) => f(snapshotsCountKey, snapshotCount)
+      case None => ()
+    }
   }
 
-  /**
-    * Model to be used to store, by block number, which block keys are no longer needed (and can potentially be deleted)
-    */
-  case class PruneCandidates(nodeKeys: Seq[ByteString])
+  private def snapshotKeysUpTo(blockNumber: BigInt, snapshotCount: BigInt): Seq[ByteString] =
+    (BigInt(0) to snapshotCount).map(snapshotIndex => getSnapshotKey(blockNumber, snapshotIndex))
```

Comment:
possible performance optimization: in `getSnapshotKey` we call `blockNumber.toByteArray` for every iteration, but it doesn't change at all here. ( i dont know how much perf overhead it adds, but we could `blockNumber.toByteArray` in this function and change `getSnapshotKey` signature to take block number as `Array[Byte]`)

Reply:
Good catch. I'm about to push a solution with currying. Let me know if you like it